library/alloc/src/collections/btree/borrow.rs

```diff
@@ -0,0 +1,38 @@
+use core::cell::UnsafeCell;
+
+/// Models a reborrow of some unique reference, when you know that the reborrow
+/// and all its descendants (i.e., all pointers and references derived from it)
+/// will not be used any more at some point, after which you want to use the
+/// original unique reference again.
+///
+/// The borrow checker usually handles this stacking of borrows for you, but
+/// some control flows that accomplish this stacking are too complicated for
+/// the compiler to follow. A `DormantMutRef` allows you to check borrowing
+/// yourself, while still expressing its stacked nature, and encapsulating
+/// the unsafe code needed to do this without undefined behavior.
+pub struct DormantMutRef<'a, T>(UnsafeCell<&'a mut T>);
```

Comment:
I don't think this makes any sense. `UnsafeCell` would permit you to mutate the reference (not mutate *through* it but mutate the pointer value itself) in an `&DormanMutRef`. You don't want to do that, so is is only misleading. Also you still should use raw pointers, not references.

Reply:
I don't really understand: the raw pointer seems equally mutable to me, and `UnsafeCell` slightly higher level, but I take your judgement higher than Miri's blessing.

Reply:
Not sure what you mean by "equally mutable".

`UnsafeCell` affects shared references but not mutable references. IOW, `&mut UnsafeCell<T>` is the same as `&mut T`. `UnsafeCell` is for being able to mutate data behind a shared reference. `DormanMutRef` does nothing like that, so I do not understand how it makes any sense to add `UnsafeCell` here. Where is there [interior mutability here](https://doc.rust-lang.org/reference/interior-mutability.html)? The answer is "nowhere", and this immediately implies that `UnsafeCell` is unnecessary.

> Miri's blessing

You'd get Miri's blessing here even if you made `DormantMutRef` just a newtype around `&mut T`, the `UnsafeCell` changes nothing. But that's an instance of https://github.com/rust-lang/unsafe-code-guidelines/issues/125, it's an open question what we really want here.